### PR TITLE
Decouple email servicename

### PR DESCRIPTION
**Pull request by dorianjp**
https://trello.com/c/yeyTL5es/158-decouple-email-and-service-name

- Added new page to wizard for creating a service.
- Added new item for editing email from address in settings.
- Moved validations to wizard steps in form vs. end submit validation
- Modified validations for service name
- Modified validations for email from
- New shared form validators
- Refactoring on create service wizard
  - remove hardcoding on steps
  - refactored next steps process from client-side
  - removed organisation type form hidden fields as it was not used in system
  - added api_client validations
  - new template page for wizard
  - content changes throughout to align with design
  - test fixes

### CODE REVIEW

**Comment by dorianjp** [comment]
@anikbrazeau please review the heroku deployed app in the scenarios for add service & service settings on name and email address.

**Review by anikbrazeau — COMMENTED** [review]
@dorianjp do you think these changes might be possible?

- Is it possible above the steps to add a [Back button](https://www.figma.com/file/BuZiiFaVHxTOQfkpWZmyvL/Service-onboarding-and-set-up-Embarquement-et-configuration-du-service?node-id=294%3A8) at each step?
- Could we remove 'alpha' from the example email domain and field-adjacent domain in both wizard and settings?
- Choose a logo is appearing twice, again as body text below the heading?
- Could we make the final call to action button: "Create service" / "Créer votre service" ?
- A couple sentences in French are different? 

`"Au sein..."` --> `"Au gouvernement du Canada, le nom de votre service doit être en français et en anglais. Par exemple : Renew your passport | Renouveler votre passeport"`

`"Cette adresse courriel..."` --> `"Votre adresse courriel d’envoi aide les destinataires à faire confiance à votre service. Une bonne adresse courriel est bilingue, reconnaissable et aussi concise que possible."`

- in settings, the @ symbol shows up as a dot (.)
- the content changes in settings pages don't seem to show up for me?

**Comment by dorianjp** [comment]
> @dorianjp do you think these changes might be possible?
> 
> * Is it possible above the steps to add a [Back button]
@anikbrazeau 

You know I missed this. I can look into it, but it will require some re-work on the wizard process.

(https://www.figma.com/file/BuZiiFaVHxTOQfkpWZmyvL/Service-onboarding-and-set-up-Embarquement-et-configuration-du-service?node-id=294%3A8) at each step?
> * Could we remove 'alpha' from the example email domain and field-adjacent domain in both wizard and settings?

Alpha comes from the domain settings, this is dynamic and changes with the environment, much like the sending domain itself when notifications are sent, so production will reflect the right domain.

> * Choose a logo is appearing twice, again as body text below the heading?

Thanks, will check

> * Could we make the final call to action button: "Create service" / "Créer votre service" ?

+1

> * A couple sentences in French are different?
> 
> `"Au sein..."` --> `"Au gouvernement du Canada, le nom de votre service doit être en français et en anglais. Par exemple : Renew your passport | Renouveler votre passeport"`
> 
> `"Cette adresse courriel..."` --> `"Votre adresse courriel d’envoi aide les destinataires à faire confiance à votre service. Une bonne adresse courriel est bilingue, reconnaissable et aussi concise que possible."`

Was this changed in figma? I copied it from there in December. I can copy the new verbage if this is the final set. Are there any other changes in figma I'll need to put in?
> 
> * in settings, the @ symbol shows up as a dot (.)

I am not seeing this, can you link to an example service?

> * the content changes in settings pages don't seem to show up for me?

Which content changes? I never changed the change your service name content, so I see that is different and can update them no problem, are there others?

**Comment by anikbrazeau** [comment]
@dorianjp 

This all sounds good! If we can get the back button in it would be great... especially if people change their mind about service name once they get to sending email. Or we could also add it in after in a separate issue?

For the French it may have been reviewed by our translator Elise after the initial commit...? I think these two sentences are really all that were different between [Figma](https://www.figma.com/file/BuZiiFaVHxTOQfkpWZmyvL/Service-onboarding-and-set-up-%7C-Embarquement-et-configuration-du-service?node-id=294%3A8) and the heroku deployment.

On name page:
 "Au sein..." --> "Au gouvernement du Canada, le nom de votre service doit être en français et en anglais. Par exemple : Renew your passport | Renouveler votre passeport"

On email page:
"Cette adresse courriel..." --> "Votre adresse courriel d’envoi aide les destinataires à faire confiance à votre service. Une bonne adresse courriel est bilingue, reconnaissable et aussi concise que possible."

**Comment by dorianjp** [comment]
[Decouple Email and Service Name](https://trello.com/c/yeyTL5es/158-decouple-email-and-service-name)

**Comment by dorianjp** [comment]
[Decouple Email and Service Name](https://trello.com/c/yeyTL5es/158-decouple-email-and-service-name)

**Comment by dorianjp** [comment]
[Decouple Email and Service Name](https://trello.com/c/yeyTL5es/158-decouple-email-and-service-name)

**Review by anikbrazeau — APPROVED** [review]
Heroku deployment looks good on admin side 👍

File: app/templates/partials/add-service/step-create-service.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+<h2 class="heading-medium">{{ _('What’s your service called?') }}</h2>
+<p>
+  {{ _('A good service name is trustworthy. It uses words people who use the service recognize.') }}
+</p>
+
+<div class="form-wrap">
+  {% call form_wrapper() %}
+    {% set hint_txt = _('Up to 255 characters') %}
+
+    {{ textbox(form.name, hint=hint_txt) }}
```
Comment (AntoineAugusti) [inline app/templates/partials/add-service/step-create-service.html:23]: Wondering if you want the input to be full size. The current size may be a bit small.

Screen Shot 2021-01-11 at 15 33 14

Comment (AntoineAugusti) [reply app/templates/partials/add-service/step-create-service.html:23]: Could we leverage the HTML5 `maxlength` attribute on this input to have some browser validation?

File: app/templates/partials/add-service/step-create-service.html [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<p>{{ _('GC Notify displays your service name:') }}</p>
+<p>
+<ul class="list-bullet">
+  <li>{{ _('as the ‘from’ address on emails you send') }}</li>
+  <li>{{ _('at the start of every text message') }}</li>
+</ul>
+</p>
+
+<p>{{ _('In the Government of Canada, your service name must include both English and French.') }}
+<br/>
```
Comment (AntoineAugusti) [inline app/templates/partials/add-service/step-create-service.html:10]: May want to duplicate the line break
Screen Shot 2021-01-11 at 15 34 18

File: app/templates/partials/add-service/step-choose-email.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<p>{{ _('Your sending email address helps to build trust with recipients. A good email address is bilingual, recognizable, and as concise as possible.') }}</p>
```
Comment (AntoineAugusti) [inline app/templates/partials/add-service/step-choose-email.html:1]: Something to discuss with @anikbrazeau @ioanacontu: it seems we give the impression that people are forced to use the usual domain notification.canada.ca but we support sending from custom domains (if they are willing to do the required configuration on their side).

Is it something we want to advertise here or is it just confusing?

Screen Shot 2021-01-11 at 15 35 10

File: app/templates/partials/add-service/step-choose-logo.html [hunk 4]
```diff
@@ -1,24 +1,13 @@
-{% set header_txt = _('Choose a logo for your service') %}
-{{ page_header(header_txt) }}
-
-<p>{{ _('Choose a logo for your service') }}</p>
-
 <p>{{ _('The default logo use in email messages is the English Government of Canada signature.') }}</p>
```
Comment (AntoineAugusti) [inline app/templates/partials/add-service/step-choose-logo.html:1]: Wondering if
> The default logo use in email messages is the English Government of Canada signature.

is necessary as you're offering me the chance to pick a convenient default here
Screen Shot 2021-01-11 at 15 37 21

File: app/translations/csv/fr.csv [hunk 5]
```diff
@@ -1249,6 +1248,20 @@
 "Want to know more? <a href='{}' class='{}'>Get in touch</a>","Vous voulez en savoir plus? <a href='{}' class='{}'>Communiquez avec nous</a>"
 "Please","Veuillez"
 "contact us","nous joindre"
+"Sending email address","Adresse courriel d’envoi"
+"Change your sending email address","Créez votre adresse courriel d’envoi"
```
Comment (AntoineAugusti) [inline app/translations/csv/fr.csv:1252]:
```suggestion
"Change your sending email address","Changez votre adresse courriel d’envoi"
```

File: app/utils.py [hunk 6]
```diff
@@ -260,6 +260,10 @@ def email_safe(string, whitespace='.'):
     return string.strip('.')
 
 
+def email_safe_name(string):
```
Comment (AntoineAugusti) [inline app/utils.py:263]: There is already a method `email_safe` just before, why did you add this? Is this enough to test that a local part is valid? Maybe delegate to `email_safe` by adding a dummy domain and add tests for this new method.

File: app/templates/views/add-service.html [hunk 7]
```diff
@@ -20,11 +27,7 @@
 
 <div class="grid-row contain-floats">
   <div class="md:w-2/3 float-left py-0 px-0 px-gutterHalf box-border">
-    {% if current_step == "choose_service_name" %}
-      {% include "partials/add-service/step1-create-service.html" %}
-    {% elif current_step == "choose_logo" %}
-      {% include "partials/add-service/step2-choose-logo.html" %}
-    {% endif %}
+    {% include tmpl %}
```
Comment (AntoineAugusti) [inline app/templates/views/add-service.html:30]: Nice

File: app/main/views/add_service.py [hunk 8]
```diff
@@ -88,102 +109,81 @@ def _create_service(service_name: str, organisation_type: str, email_from: str,
...
+        step_max=len(WIZARD_ORDER),
+        tmpl=WIZARD_DICT[current_step]['tmpl'],
+        back_link=back_link
     )
 
 
 @main.route("/add-service", methods=['GET', 'POST'])
 @user_is_logged_in
 @user_is_gov_user
 def add_service():
```
Comment (AntoineAugusti) [inline app/main/views/add_service.py:138]: The form logic is really intense to read. Did you try to refactor it a bit, make it a bit easier to read? It works fine, the UI is great but I have a hard wrapping my head around the following lines.

Maybe the contact page has good things to steal from

File: app/main/validators.py [hunk 9]
```diff
@@ -135,3 +135,32 @@ def __init__(self, message="Can't start with 00"):
...
+    service_id = getattr(form, 'service_id', current_app.config['NOTIFY_BAD_FILLER_UUID'])
+    unique_name = service_api_client.is_service_email_from_unique(
+        service_id,
+        field.data
+    )
+    if not unique_name:
+        raise ValidationError(_l("This email address is already in use"))
+
+
+def validate_name(form, field):
```
Comment (AntoineAugusti) [inline app/main/validators.py:155]: Rename to `validate_service_name`? Name is a bit too generic I think, I think about a first name when reading this.

File: app/config.py [hunk 10]
```diff
@@ -34,6 +34,7 @@ class Config(object):
     NOTIFY_SERVICE_ID = 'd6aa2c68-a2d9-4437-ab19-3ae8eb202553'
     NOTIFY_USER_ID = '6af522d0-2915-4e52-83a3-3690455a5fe6'
     NOTIFY_TEMPLATE_PREFILL_SERVICE_ID = '93305b36-b0a0-4a34-9ab2-c1b7bb5ca489'
+    NOTIFY_BAD_FILLER_UUID = '00000000-0000-0000-0000-000000000000'
```
Comment (AntoineAugusti) [inline app/config.py:37]: I've seen where this is used. WDYT about moving this to the service creation logic and not in the config? I guess you also generate a random UUID if you feel like it.

File: tests/app/main/views/test_service_settings.py [hunk 11]
```diff
@@ -248,7 +256,10 @@ def test_should_show_overview_for_service_with_more_things_set(
     ))
     page = BeautifulSoup(response.data.decode('utf-8'), 'html.parser')
     for index, row in enumerate(expected_rows):
-        assert row == " ".join(page.find_all('tr')[index + 1].text.split())
+        formatted_row = row.format(sending_domain=os.environ.get(
+            'SENDING_DOMAIN', 'notification.alpha.canada.ca'
+        ))
```
Comment (AntoineAugusti) [inline tests/app/main/views/test_service_settings.py:261]: You can move the fallback to the config (or update CI/local envs) if you feel like it

**Review by AntoineAugusti — COMMENTED** [review]
Good work!

Made a few suggestions/questions. Some are about the design, @anikbrazeau and @ioanacontu may help but this is not blocking.

File: app/utils.py (hunk 6 above)
Comment (dorianjp) [reply app/utils.py:263]: @AntoineAugusti it's to prevent double quotes from polluting the sender smtp header as the celery jobs break unless we do something with double quotes like encode etc. for the final SMTP transfer since they're handled raw. In the end, we don't need double quotes for service name, and this function is a good validation bottleneck for name formatting.

File: tests/app/main/views/test_service_settings.py (hunk 11 above)
Comment (dorianjp) [reply tests/app/main/views/test_service_settings.py:261]: @AntoineAugusti , You know I tried injecting the os environ SENDING_DOMAIN through the mock service in the dictionary object's pre-existing sending_domain field, and it was never carried over. I may have missed something, but this worked for now.

File: app/main/views/add_service.py (hunk 8 above)
Comment (dorianjp) [reply app/main/views/add_service.py:138]: Added some comments to enhance readability. The contact page is missing some edge cases covered here on out-of-order form submission.

File: app/templates/partials/add-service/step-create-service.html (hunk 1 above)
Comment (dorianjp) [reply app/templates/partials/add-service/step-create-service.html]: +1, Added in latest commit

File: app/templates/partials/add-service/step-create-service.html (hunk 2 above)
Comment (dorianjp) [reply app/templates/partials/add-service/step-create-service.html]: +1, Added in latest commit

File: app/main/validators.py (hunk 9 above)
Comment (dorianjp) [reply app/main/validators.py]: +1, Added in latest commit

File: app/config.py (hunk 10 above)
Comment (dorianjp) [reply app/config.py:37]: I put it there based on similar 'test-y' type data in the same block within the app. I felt the zeros are a guaranteed non-collision in the same vein as the other data. I'm not crazy about using filler data, but felt it reduced complexity of the form validation in cases where a service_id doesn't exist yet we can leverage the same validation routine in forms and API.

File: app/templates/partials/add-service/step-choose-email.html (hunk 3 above)
Comment (anikbrazeau) [reply app/templates/partials/add-service/step-choose-email.html:1]: Good point, noted and will consider revisiting in next iteration!

File: app/templates/partials/add-service/step-choose-logo.html (hunk 4 above)
Comment (anikbrazeau) [reply app/templates/partials/add-service/step-choose-logo.html:1]: You're right, it's redundant now, I'll make a separate issue to remove it :)
